core/application/impl/app_config_impl.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_APP_CONFIG_IMPL_HPP
+#define KAGOME_APP_CONFIG_IMPL_HPP
+
+#include "application/app_config.hpp"
+
+#include <rapidjson/document.h>
+#include <cstdio>
+#include <memory>
+
+#include "common/logger.hpp"
+
+#ifdef DECLARE_PROPERTY
+#error DECLARE_PROPERTY already defined!
+#endif  // DECLARE_PROPERTY
+#define DECLARE_PROPERTY(T, N)                                                 \
+ private:                                                                      \
+  T N##_;                                                                      \
+                                                                               \
+ public:                                                                       \
+  std::conditional<std::is_trivial<T>::value && (sizeof(T) <= sizeof(size_t)), \
+                   T,                                                          \
+                   const T &>::type                                            \
+  N() const override {                                                         \
+    return N##_;                                                               \
+  }
+
+namespace kagome::application {
+
+  class AppConfigurationImpl final : public AppConfiguration {
+    static void file_deleter(std::FILE *f);
+    using FilePtr = std::unique_ptr<std::FILE, decltype(&file_deleter)>;
```

Comment:
You might to simplify that:
```suggestion
    using FilePtr = std::unique_ptr<std::FILE, std::fclose>;
```